app/src/main/java/com/victor/adasfeed/ui/recyclerview/adapter/PostAdapter.kt

```diff
@@ -0,0 +1,88 @@
+package com.victor.adasfeed.ui.recyclerview.adapter
+
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.ImageView
+import android.widget.TextView
+import android.widget.Toast
+import androidx.recyclerview.widget.RecyclerView
+import androidx.recyclerview.widget.RecyclerView.ViewHolder
+import com.google.android.material.snackbar.Snackbar
+import com.victor.adasfeed.R
+import com.victor.adasfeed.mock.Faker
+import com.victor.adasfeed.model.Post
+import com.victor.adasfeed.repository.LikedPostsRepo
+
+class PostAdapter(private val postList: MutableList<Post> = Faker.makePostList()) :
+    RecyclerView.Adapter<PostAdapter.PostViewHolder>() {
+
+    class PostViewHolder(view: View) : ViewHolder(view) {
+        val tvUsername: TextView = view.findViewById(R.id.textUserName)
+        val tvDescription: TextView = view.findViewById(R.id.textPostDescription)
+        val ivUser: ImageView = view.findViewById(R.id.imageUser)
+        val ivPost: ImageView = view.findViewById(R.id.imagePost)
+        val ivHeart: ImageView = view.findViewById(R.id.imageLiked)
+        val ivBookmark: ImageView = view.findViewById(R.id.imageFavorite)
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PostViewHolder {
+        val postView =
+            LayoutInflater.from(parent.context).inflate(R.layout.post_item, parent, false)
+        return PostViewHolder(postView)
+    }
+
+    override fun getItemCount() = postList.size
+
+    override fun onBindViewHolder(holder: PostViewHolder, position: Int) {
+        holder.apply {
+            val p = postList[position]
+
+            tvUsername.text = p.userName
+            tvDescription.text = p.description
+            ivPost.setImageResource(p.imagePost)
+            ivUser.setImageResource(p.imageUser)
+
+            ivHeart.setOnClickListener {
+                p.isLiked = !p.isLiked
+
+                if (p.isLiked) {
+                    ivHeart.setImageResource(R.drawable.favorite_filled)
+                    LikedPostsRepo.add(p)
+                } else {
+                    ivHeart.setImageResource(R.drawable.favorite)
+                    LikedPostsRepo.remove(p)
+                }
+
+                Toast.makeText(ivHeart.context, LikedPostsRepo.toString(), Toast.LENGTH_SHORT).show()
+            }
+
+            ivPost.setOnLongClickListener {
+                removePost(position)
+
+                Snackbar.make(it, "Post deletado", 5000).apply{
+                    setAction("desfazer") {
+                        insertPost(p, position)
+                    }
+                }.show()
+
+                true
+            }
+        }
+    }
+
+    fun addPost(p: Post) {
+        postList.add(p)
+        notifyItemInserted(postList.indexOf(p))
```

Comment:
aqui poderia estar sendo utilizado o DiffUtil para atualizar apenas os itens que mudaram.

Reply:
Esse meu comentário conta para todos os comentários sobre DiffUtil nesse PR.

Oi oi, Raiara, boa noite. Pelo que eu entendo do DiffUtil ele calcularia as mudanças mínimas pra transformar uma lista velha em uma nova. Isso seria feito em toda lista, então verificaria tudo, daí o que tivesse sofrido update seria atualizado. Acaba ficando uma complexidade maior nesse caso desde que apenas adicionamos um post.

Complexidade de acordo com documentação do DiffUtil:

"This algorithm is optimized for space and uses O(N) space to find the minimal number of addition and removal operations between the two lists. It has O(N + D^2) expected time performance where D is the length of the edit script."

Como nesse caso a gente só atualiza um post acho q não há necessidade e é uma operação menos complexa.

Tá aqui o porquê do notifyItemInserted ser menos complexa:

"Notify any registered observers that the item reflected at position has been newly inserted. The item previously at position is now at position position + 1.

This is a structural change event. Representations of other existing items in the data set are still considered up to date and will not be rebound, though their positions may be altered."

Os itens que não sofreram alteração não são nem verificados, não precisamos verificar toda lista, apenas quem foi adicionado.


O que você acha? Talvez eu tenha deixado algo passar do DiffUtil, se puder explicar mais explicitamente o porque do DiffUtil ser mais performático nesse caso me ajudaria muito. Desde já agradeço :)

Reply:
Olá! Obrigada por compartilhar sua opinião e perspectiva sobre a utilização do DiffUtil neste caso específico.

O notify funciona, mas a classe DiffUtil já foi definida no adapter, já que foi definido acho que faria sentido padronizar e utilizá-la

E como você mesmo citou ali no texto, é performático, a complexidade de tempo é O(N) e isso é a forma mais rápida então em termos de escalabilidade utilizar o diffutil não vai onerar o desempenho de forma alguma

Mas também meu comentário é só uma sugestão. De qualquer forma, acredito que ambas as opções podem ser válidas, cabe a você escolher o que for melhor :)